framework/rediscache.py

```diff
@@ -0,0 +1,109 @@
+
+
+# -*- coding: utf-8 -*-
+# Copyright 2020 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import logging
+
+from google.cloud import ndb
+import redis
+import fakeredis
+
+
+UNIT_TEST_MODE = os.environ['SERVER_SOFTWARE'].startswith('test')
+if not UNIT_TEST_MODE:
+  # Create a Redis client.
+  redis_host = os.environ.get('REDISHOST', 'localhost')
+  redis_port = int(os.environ.get('REDISPORT', 6379))
+  redis_client = redis.Redis(host=redis_host, port=redis_port, decode_responses=True)
+else:
+  redis_client = fakeredis.FakeStrictRedis(decode_responses=True)
+
+
+def set(key, value, time=None):
+  """
+  Redis SET sets the str key, value pair, https://redis.io/commands/set/; if
+  ``key`` already holds a value, it is overwritten.
+
+  ``time`` sets the expire time for this key, in seconds.
+  """
+  if not isinstance(value, str):
+    logging.info(
+        'value %s is not an instance of str, abort set caching', value)
+    return
+
+  if time:
+    redis_client.set(key, value, ex=time)
+  else:
+    redis_client.set(key, value)
+
+
+def get(key):
+  """
+  Redis GET gets the value of key. Return nil if ``key`` does not
+  exist; return an error if the value returned is not a str.
+  """
+  return redis_client.get(key)
+
+
+def get_multi(keys):
+  """Return the values of all given keys."""
+  vals = redis_client.mget(keys)
+  return dict(zip(keys, vals))
+
+
+def set_multi(entries, time=None):
+  """
+  Set the given keys to their respective values.
+
+  ``time`` sets the expire time for this key, in seconds.
+  """
+  data_entries = {}
+  for key in entries:
+    if not isinstance(entries[key], str):
+      logging.info(
+          'value %s is not an instance of str, skipping this cache', entries[key])
+      continue
+    data_entries[key] = entries[key]
+
+  if not time:
+    # https://redis.io/commands/mset/.
+    redis_client.mset(data_entries)
+    return
+
+  for key in data_entries:
+    set(key, data_entries[key], time)
+
+
+def delete(key):
+  """Redis DEL removes the value to the key, https://redis.io/commands/del/."""
+  redis_client.delete(key)
+
+
+def flushall():
+  """Delete all the keys in Redis, https://redis.io/commands/flushall/."""
+  redis_client.flushall()
+
+def serialize_non_str(data):
```

Comment:
Most Redis operations only accept `str` and `binary` as value, per https://redis.io/commands/set/. I chose `str` because it is more human readable when we look at cache in the memorystore. For that reason, we have to serialize all `non-str` types into JSON. ndb entities are trickier because they are objects and we cannot directly covert them into JSON

Reply:
`class AppUser(core_models.DictModel):` is an instance of how I decided to store NDB entities. We have to explicitly convert non-primitive types into a string

Reply:
Also the reason why I created serialize and deserialize is that user should made aware of the type of values in Redis - it need to be serializable so that it can be cached to Redis

Reply:
I think a good metric to guide code design decisions is simply the number of source code lines needed to implement it one way or another.   If two approaches both use the same set of language features, I usually favor the design that has fewer total lines of code.   I think that we have a lot of code that will call rediscache, so making those calls be a single line each rather than two lines will make it a lower total number of lines, even if it adds a few lines inside rediscache.

In rediscache.py, let's try giving redis the output of `pickle.dumps()` and then when we get values back, use `pickle.loads()` to deserialize the values.   This should be safe because this is only data that our own app reads and writes.  It will make the values stored in redis non-human readable, but we can debug by looking at values on the way in or out in rediscache.py.   `pickle` will be better than using json strings because pickled objects have the name of the python class, so it should handle all ndb.Model subclasses.  You can pass any kind of python object to `pickle.dumps()`, so that should work in any existing cases where we cache a string, number, list, or dict, just as well as the model instances.

Reply:
Done